python/cudf/cudf/core/join/join.py

```diff
@@ -0,0 +1,431 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.
+
+import warnings
+
+import numpy as np
+import pandas as pd
+import itertools
+
+import cudf._lib as libcudf
+from cudf.utils.dtypes import is_categorical_dtype, is_datetime_dtype
+import cudf
+
+class Merge(object):
+    def __init__(
+        self,
+        lhs,
+        rhs,
+        on,
+        left_on,
+        right_on,
+        left_index,
+        right_index,
+        how,
+        sort,
+        lsuffix,
+        rsuffix,
+        method,
+        indicator,
+        suffixes,
+    ):
+        self.lhs = lhs
+        self.rhs = rhs
+        self.left_index = left_index
+        self.right_index = right_index
+        self.method = method
+        self.sort = sort
+
+        # check that the merge is valid
+
+        self.validate_merge_cfg(
+            lhs,
+            rhs,
+            on,
+            left_on,
+            right_on,
+            left_index,
+            right_index,
+            how,
+            lsuffix,
+            rsuffix,
+        )
+        self.how = how
+        self.preprocess_merge_params(on, left_on, right_on, lsuffix, rsuffix)
+
+    def perform_merge(self):
+        output_dtypes = self.compute_output_dtypes()
+        self.typecast_input_to_libcudf()
+        libcudf_result = libcudf.join.join(
+            self.lhs,
+            self.rhs,
+            self.left_on,
+            self.right_on,
+            self.how,
+            self.method,
+            left_index=self.left_index,
+            right_index=self.right_index,
+        )
+        result = self.typecast_libcudf_to_output(libcudf_result, output_dtypes)
+        return result
+
+    def preprocess_merge_params(self, on, left_on, right_on, lsuffix, rsuffix):
+        if on:
+            on = [on] if isinstance(on, str) else list(on)
+            left_on = right_on = on
+        else:
+            on = []
+        if left_on:
+            left_on = [left_on] if isinstance(left_on, str) else list(left_on)
+        else:
+            left_on = []
+        if right_on:
+            right_on = (
+                [right_on] if isinstance(right_on, str) else list(right_on)
+            )
+        else:
+            right_on = []
```

Comment:
What's the priority here? If someone sets `on`, it will then go down the `left_on` and `right_on` branches as well and make copies of the list unnecessarily. Should those be `elif`?

Reply:
I did away with this and reworked these conditions.